src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfAnalyzer.cs

```diff
@@ -0,0 +1,184 @@
+
+                var stringArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType));
+                var charArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType));
+                var stringAndComparisonTypeArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charAndComparisonTypeArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+
+                // Check that the contains methods that take 2 parameters exist
+                // string.Contains(char) is also .NETStandard2.1+
+                var stringContainsMethods = stringType
+                    .GetMembers("Contains")
+                    .OfType<IMethodSymbol>()
+                    .WhereAsArray(s =>
+                        s.Parameters.Length <= 2);
+                var stringAndComparisonTypeArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charAndComparisonTypeArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType));
+                if (stringAndComparisonTypeArgumentContainsMethod == null ||
+                    charAndComparisonTypeArgumentContainsMethod == null ||
+                    charArgumentContainsMethod == null)
+                {
+                    return;
+                }
+
+                // Roslyn doesn't yet support "FindAllReferences" at a file/block level. So instead, find references to local int variables in this block.
+                context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+
+                void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+                {
+                    if (!(context.OwningSymbol is IMethodSymbol method))
+                    {
+                        return;
+                    }
+
+                    PooledConcurrentSet<ISymbol> localsToBailOut = PooledConcurrentSet<ISymbol>.GetInstance();
```

Comment:
I think it would be good to add an explanatory comment about the algorithm and data structures used here so it is easier to understand for someone else in future.